File: JS/LeetC/queue/circular-queue.js
// Implement a queue using "circular" array

class MyCircularQueue {
  constructor(k = 5) {
    this.size = k;
    this.data = [];
    this.head = null;
    this.tail = null;
  }

  // Gets the front item from the queue. If the queue is empty, return -1
  Front() {
    if (!this.isEmpty()) {
      return this.data[this.head];
    }

    return -1;
  }

  // Gets the last item from the queue. If the queue is empty, return -1
  Rear() {
    if (!this.isEmpty()) {
      return this.data[this.tail];
    }

    return -1;
  }

  // enQueue :: int -> bool
  enQueue(val) {
    if (this.isFull()) {
      return false;
    }

    if (this.isEmpty()) {
      this.head = 0;
      this.tail = 0;
    } else {
      this.tail = (this.tail + 1) % this.size;
    }

    this.data[this.tail] = val;
    return true;
  }

  // deQueue :: void -> bool
  deQueue() {
    if (this.isEmpty()) {
      return false;
    }

    if (this.tail === this.head) {
      this.data[this.tail] = undefined;
      this.tail = null;
      this.head = null;
      return true;
    }

    this.head = (this.head + 1) % this.size;
    return true;
  }

  // isEmpty : void -> bool
  isEmpty() {
    return this.head === null || this.tail === null;
  }

  isFull() {
    return (
      this.head !== null &&
      this.tail !== null &&
      ((this.head === 0 && this.tail === this.size - 1) ||
        (this.head > this.tail && this.head - this.tail === 1))
    );
  }
}

module.exports = MyCircularQueue;

File: python/leetcode/queues/circular_queue.py
class MyCircularQueue:
    def __init__(self, k=10):
        self.size = k
        self.head = None
        self.tail = None
        self.data = [None] * k

    def enQueue(self, value):
        if self.isFull():
            return False

        if self.isEmpty():
            self.head = 0

        self.tail = 0 if self.isEmpty() else ((self.tail + 1) % self.size)
        self.data[self.tail] = value
        return True

    def deQueue(self):
        if self.isEmpty():
            return False

        self.data[self.head] = None

        self.head = (self.head + 1) % self.size

        if self.data[self.head] is None:
            self.head = None
            self.tail = None

        return True

    def Front(self):
        if self.isEmpty():
            return -1
        return self.data[self.head]

    def Rear(self):
        if self.isEmpty():
            return -1
        return self.data[self.tail]

    def isEmpty(self) -> bool:
        return self.head is None or self.tail is None

    def isFull(self) -> bool:
        if self.isEmpty():
            return False

        return (self.tail + 1) % self.size == self.head


if __name__ == "__main__":
    c = MyCircularQueue(5)
    print(c.enQueue(1))
    print(c.data)
    print(c.enQueue(2))
    print(c.data)
    print(c.enQueue(3))
    print(c.data)
    print(c.enQueue(4))
    print(c.data)
    print(c.enQueue(5))
    print(c.data)
    print('head: ', c.head)
    print('tail: ', c.tail)
    print('is full: ', c.isFull())
    print('is empty: ', c.isEmpty())

    print('----------------------')
    print(c.deQueue())
    print(c.data)
    print(c.deQueue())
    print(c.data)
    print(c.deQueue())
    print(c.data)
    print(c.deQueue())
    print(c.data)
    print(c.deQueue())
    print(c.data)
    print('is full: ', c.isFull())
    print('is empty: ', c.isEmpty())
